### PR TITLE
feat(guild): add `bulk_ban` function

### PR DESCRIPTION
**Pull request by spifory**
## Summary

Implements discord/discord-api-docs#6720 by adding a `bulk_ban` function to `Guild`

Tested with: https://paste.nextcord.dev/?id=1729350208916274
image
(component gets disabled so the users disappear from the select)

## This is a **Code Change**

- [x] I have tested my changes.
- [x] I have updated the documentation to reflect the changes.
- [x] I have run `task pyright` and fixed the relevant issues.

### CODE REVIEW

**Review by teaishealthy — CHANGES_REQUESTED** [review]
code looks good

> If all users "failed" you get an error response instead

but i feel like this is a bit too unexpected, should go under a `Raises: HTTPException: ...` (?)

File: nextcord/guild.py [hunk 1]
```diff
@@ -3079,6 +3080,41 @@ async def ban(
 
         await self._state.http.ban(user.id, self.id, delete_message_seconds, reason=reason)
 
+    async def bulk_ban(self, users: List[Snowflake], *, delete_message_seconds: int = 0) -> BulkBan:
+        """|coro|
+
+        Bans users from a guild in bulk. This has a limit of 200 users
```
Comment (EmreTech) [inline nextcord/guild.py:3086]:
```suggestion
        Bans a list of users. This is similar to :meth:`Guild.ban` except it bulk bans multiple users. 
        
        This has a limit of 200 users.
```
Changed to be phrased more like `Thread.delete_messages`.

File: nextcord/http.py [hunk 2]
```diff
@@ -877,6 +877,17 @@ def ban(
 
         return self.request(r, params=params, reason=reason)
 
+    def bulk_ban(
```
Comment (EmreTech) [inline nextcord/http.py:880]: This supports audit log reasons, according to the Discord Docs PR.

File: nextcord/guild.py [hunk 3]
```diff
@@ -3079,6 +3080,41 @@ async def ban(
...
+
+        The list of users must meet the :class:`abc.Snowflake` abc.
```
Comment (EmreTech) [inline nextcord/guild.py:3088]:
```suggestion
```
This part is pretty self-explanatory, and most users would provide the `User` objects anyway.

File: nextcord/guild.py [hunk 4]
```diff
@@ -3079,6 +3080,41 @@ async def ban(
...
+
+        You must have the :attr:`~Permissions.ban_members` and :attr:`~Permissions.manage_guild` permissions to
+        do this.
+
+        Parameters
+        ----------
+        users: :class:`abc.Snowflake`
```
Comment (EmreTech) [inline nextcord/guild.py:3095]:
```suggestion
        users: List[:class:`abc.Snowflake`]
```

File: nextcord/guild.py [hunk 5]
```diff
@@ -3079,6 +3080,41 @@ async def ban(
...
+            The users to ban from a guild.
+        delete_message_seconds: :class:`int`
```
Comment (EmreTech) [inline nextcord/guild.py:3097]: Should we make this optional to match the behavior of `Guild.ban`?

File: nextcord/guild.py [hunk 6]
```diff
@@ -3079,6 +3080,41 @@ async def ban(
...
+            The number of seconds worth of messages to delete from these users.
+            Can range from 0 to 604800 seconds (7 days).
+
+            Defaults to 0.
+
```
Comment (EmreTech) [inline nextcord/guild.py:3102]: The errors that are thrown from this function should be documented here. `Forbidden` should be thrown if there aren't sufficient permissions, and `HTTPException` should be thrown if no users were banned.

File: nextcord/bans.py [hunk 7]
```diff
@@ -13,3 +15,8 @@
 class BanEntry(NamedTuple):
     reason: Optional[str]
     user: "User"
+
+
+class BulkBan(NamedTuple):
```
Comment (EmreTech) [inline nextcord/bans.py:20]: Make sure to add this to `api.rst`.

**Review by EmreTech — CHANGES_REQUESTED** [review]
Code looks good otherwise!

**Comment by EmreTech** [comment]
Related Discord Docs PR was merged.

File: nextcord/guild.py [hunk 8]
```diff
@@ -3079,6 +3080,57 @@ async def ban(
 
         await self._state.http.ban(user.id, self.id, delete_message_seconds, reason=reason)
 
+    async def bulk_ban(
+        self,
+        users: List[Snowflake],
+        *,
+        delete_message_seconds: Optional[int] = None,
+        reason: Optional[str] = None,
```
Comment (EmreTech) [inline nextcord/guild.py:3088]: You forgot to document this parameter.

**Review by EmreTech — COMMENTED** [review]
Looks good otherwise!

File: nextcord/guild.py [hunk 9]
```diff
@@ -3079,6 +3080,59 @@ async def ban(
...
+
+        Parameters
+        ----------
+        users: List[:class:`abc.Snowflake`]
+            The users to ban from a guild.
+        delete_message_seconds: Optional[:class:`int`]
+            The number of seconds worth of messages to delete from these users.
+            Can range from 0 to 604800 seconds (7 days).
+
+            Defaults to 0.
```
Comment (teaishealthy) [inline nextcord/guild.py:3105]: why does the default differ from the the Guild.ban default

File: nextcord/guild.py [hunk 10]
```diff
@@ -3079,6 +3080,59 @@ async def ban(
...
+        :class:`BulkBan`
+            The failed and banned users in the form of :class:`Object`.
+        """
+
+        if delete_message_seconds is None:
+            delete_message_seconds = 0
+
+        data = await self._state.http.bulk_ban(
+            self.id,
+            [u.id for u in users],
```
Comment (ooliver1) [inline nextcord/guild.py:3126]: Using named keyword parameters makes more semantic sense than positional parameters.

File: nextcord/guild.py [hunk 11]
```diff
@@ -3079,6 +3080,59 @@ async def ban(
...
+        users: List[:class:`abc.Snowflake`]
+            The users to ban from a guild.
+        delete_message_seconds: Optional[:class:`int`]
+            The number of seconds worth of messages to delete from these users.
+            Can range from 0 to 604800 seconds (7 days).
+
+            Defaults to 0.
+        reason: Optional[:class:`str`]
+            The reason these users got banned.
+        Raises
```
Comment (ooliver1) [inline nextcord/guild.py:3108]:
```suggestion

        Raises
```

**Review by ooliver1 — CHANGES_REQUESTED** [review]
You can add versionadded 3.0 too since this should ideally be in for that.

File: nextcord/http.py [hunk 12]
```diff
@@ -1862,6 +1862,21 @@ def ban(
             retry_request=retry_request,
         )
 
+    def bulk_ban(
+        self,
+        guild_id: Snowflake,
+        user_ids: List[Snowflake],
```
Comment (alentoghostflame) [inline nextcord/http.py:1868]: The positional args for .ban are  `user_id, guild_id` but .bulk_ban has them reversed (`guild_id, user_ids`)? Maybe I'm being nitpicky, but it feels inconsistent?

Comment (spifory) [reply nextcord/http.py:1868]: Good point, I (most likely) didn't notice it when the code

**Review by Skelmis — COMMENTED** [review]
Outside of alento's comment, lgtm

**Review by EmmmaTech — APPROVED** [review]
looks good :D
haven't tested

**Review by Skelmis — APPROVED** [review]

**Comment by Skelmis** [comment]
@ooliver1 @teaishealthy this pr up to y'all

File: docs/api.rst [hunk 13]
```diff
@@ -2374,6 +2374,22 @@ Guild
 
         :type: :class:`User`
 
+.. class:: BulkBan
+
+    A namedtuple which represents a ban returned from :meth:`~Guild.bulk_ban`.
+
```
Comment (ooliver1) [inline docs/api.rst:2380]:
```suggestion

    .. versionadded:: 3.0

```

File: nextcord/http.py [hunk 14]
```diff
@@ -1856,6 +1856,21 @@ def ban(
...
+        reason: Optional[str] = None,
+        *,
+        auth: Optional[str] = MISSING,
+        retry_request: bool = True,
+    ) -> Response[guild.BulkBan]:
+        r = Route("POST", "/guilds/{guild_id}/bulk-ban", guild_id=guild_id)
+        params = {"users_ids": user_ids, "delete_message_seconds": delete_message_seconds}
+
+        return self.request(r, params=params, reason=reason, auth=auth, retry_request=retry_request)
+
```
Comment (ooliver1) [inline nextcord/http.py:1873]:
```suggestion
        data = {"user_ids": user_ids, "delete_message_seconds": delete_message_seconds}

        return self.request(r, json=data, reason=reason, auth=auth, retry_request=retry_request)
```
These parameters must be passed as json and the parameter name is `user_ids`, not `users_ids`.

```
nextcord.ext.commands.errors.CommandInvokeError: Command raised an exception: HTTPException: 400 Bad Request (error code: 50035): Invalid Form Body
In _errors: [{'code': 'CONTENT_TYPE_INVALID', 'message': 'Expected "Content-Type" header to be one of {\'application/json\'}.'}]
```
```
nextcord.errors.HTTPException: 400 Bad Request (error code: 50035): Invalid Form Body
In user_ids: This field is required
```

**Review by ooliver1 — CHANGES_REQUESTED** [review]
Please test all PRs in future. I used the following test code
```python
@bot.command()
async def bulk_ban(ctx: Context, users: Greedy[User], *, reason: Optional[str] = None):
    res = await ctx.guild.bulk_ban(users, reason=reason)
    await ctx.send(f"Banned {len(users)} users!")
```

File: nextcord/http.py [hunk 15]
```diff
@@ -1856,6 +1856,21 @@ def ban(
...
+        delete_message_seconds: int = 0,
+        reason: Optional[str] = None,
+        *,
+        auth: Optional[str] = MISSING,
+        retry_request: bool = True,
+    ) -> Response[guild.BulkBan]:
+        r = Route("POST", "/guilds/{guild_id}/bulk-ban", guild_id=guild_id)
+        params = {"users_ids": user_ids, "delete_message_seconds": delete_message_seconds}
+
+        return self.request(r, params=params, reason=reason, auth=auth, retry_request=retry_request)
```
Comment (ooliver1) [inline nextcord/http.py:1872]:
```suggestion
        data = {"user_ids": user_ids, "delete_message_seconds": delete_message_seconds}

        return self.request(r, json=data, reason=reason, auth=auth, retry_request=retry_request)
```
These parameters must be passed as json and the parameter name is `user_ids`, not `users_ids`.

```
nextcord.ext.commands.errors.CommandInvokeError: Command raised an exception: HTTPException: 400 Bad Request (error code: 50035): Invalid Form Body
In _errors: [{'code': 'CONTENT_TYPE_INVALID', 'message': 'Expected "Content-Type" header to be one of {\'application/json\'}.'}]
```
```
nextcord.errors.HTTPException: 400 Bad Request (error code: 50035): Invalid Form Body
In user_ids: This field is required
```

File: nextcord/http.py [hunk 16]
```diff
@@ -1856,6 +1856,21 @@ def ban(
             retry_request=retry_request,
         )
 
+    def bulk_ban(
+        self,
+        user_ids: List[Snowflake],
+        guild_id: Snowflake,
+        delete_message_seconds: int = 0,
+        reason: Optional[str] = None,
+        *,
```
Comment (ooliver1) [inline nextcord/http.py:1865]:
```suggestion
        guild_id: Snowflake,
        delete_message_seconds: int = 0,
        reason: Optional[str] = None,
        *,
```

**Review by ooliver1 — APPROVED** [review]
image
tested

**Pull request merged**